utils/appliance/__init__.py

```diff
@@ -249,37 +250,103 @@ def seal_for_templatizing(self):
+                    ip_addresses.add(ep.hostname)
+                else:
+                    ip_address = resolve_hostname(ep.hostname)
+                    if ip_address is not None:
+                        ip_addresses.add(ip_address)
+        return list(ip_addresses)
+
     @property
     def managed_providers(self):
-        """Returns a set of providers that are managed by this appliance
+        """Returns a set of provider crud objects of known providers managed by this appliance
 
-        Returns:
-            :py:class:`set` of :py:class:`str` - provider_key-s
+        Note:
+            Provider types have different ways of identification:
+              Infra = hostname / ip
+              Cloud = credentials
+              Containers = hostname / ip
+              Middleware = hostname / ip
+
+        Warning:
+            If name of the provider on the appliance differs from name in yamls, the name from the
+            appliance is used instead so that we can interact with the provider.
         """
-        ip_addresses = set([])
-
+        ems_table = self.db["ext_management_systems"]
         # Fetch all providers at once, return empty list otherwise
         try:
-            query_res = list(self._query_endpoints())
+            found_ems = list(self.db.session.query(ems_table))
         except Exception as ex:
             self.log.warning("Unable to query DB for managed providers: %s", str(ex))
             return []
 
-        for ipaddress, hostname in query_res:
-            if ipaddress is not None:
-                ip_addresses.add(ipaddress)
-            elif hostname is not None:
-                if re.match(r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$", hostname) is not None:
-                    ip_addresses.add(hostname)
-                else:
-                    ip_address = resolve_hostname(hostname)
-                    if ip_address is not None:
-                        ip_addresses.add(ip_address)
-        provider_keys = set([])
-        for provider_key, provider_data in conf.cfme_data.get("management_systems", {}).iteritems():
-            if provider_data.get("ipaddress", None) in ip_addresses:
-                provider_keys.add(provider_key)
-        return provider_keys
+        from utils.providers import list_providers
+        creds_table = self.db["authentications"]
+        prov_cruds = list_providers()
+        recognized_by_ip = ["InfraManager", "ContainerManager", "MiddlewareManager"]
+        recognized_by_creds = ["CloudManager"]
+        ignored_types = ["NetworkManager"]
+
+        found_cruds = set()
+        unrecognized_ems_names = set()
+        for ems in found_ems:
+            # Skip any provider types that we don't care about
+            if any(p_type in ems.type for p_type in ignored_types):
+                continue
+            found_crud = None
+            for prov in prov_cruds:
+                # Name check is authoritative
+                if ems.name == prov.name:
```

Comment:
What do you mean by edited on the appliance? I'm also confused when you point out cloud as if it doesn't have endpoints? RHOS clearly does and in some test envs the creds could be the same for multiple RHOS envs. How do we handle this? Ultimately, I think it we should give people an easy way to setup the provider via the automation so they don't do it manually. I now almost always do this via an ipython precisely to avoid this issue. Though I do get the sentiment, I wonder if we are mollycoddling ever so slightly.